File: backend/src/main/java/de/interaapps/pastefy/auth/strategies/oauth2/providers/DiscordOAuth2Provider.java
package de.interaapps.pastefy.auth.strategies.oauth2.providers;

import org.javawebstack.abstractdata.AbstractObject;
import org.javawebstack.abstractdata.util.QueryString;
import org.javawebstack.httpclient.HTTPClient;
import de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Profile;
import de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Provider;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;

public class DiscordOAuth2Provider implements OAuth2Provider {
    private String clientId;
    private String secret;
    private String[] scopes = {"email","identify"};
    private HTTPClient discordClient;

    public DiscordOAuth2Provider(String clientId, String secret){
        this.clientId = clientId;
        this.secret = secret;
        discordClient = new HTTPClient("https://discord.com");
    }

    public DiscordOAuth2Provider setScopes(String... scopes) {
        this.scopes = scopes;
        return this;
    }

    public OAuth2Callback callback(AbstractObject queryParameters, String callbackUrl) {
        AbstractObject abstractObject = discordClient.post("/api/oauth2/token")
                .header("User-Agent", "JWSPassportClient/1")
                .formBodyString(new QueryString()
                        .set("client_id", clientId)
                        .set("client_secret", secret)
                        .set("code", queryParameters.string("code"))
                        .set("grant_type", "authorization_code")
                        .set("redirect_uri", callbackUrl)
                        .set("scope", String.join(" ", scopes))
                        .toString()
                )
                .data().object();
        return new OAuth2Callback(this, abstractObject.string("access_token"), abstractObject.string("refresh_token"));
    }

    public String redirect(String callbackUrl) {
        try {
            return "https://discord.com/api/oauth2/authorize?response_type=code&client_id="+clientId+"&prompt=consent&scope="+ URLEncoder.encode(String.join(" ", scopes), "UTF-8")+"&redirect_uri="+URLEncoder.encode(callbackUrl, "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return "";
    }

    public static class OAuth2Callback extends de.interaapps.pastefy.auth.strategies.oauth2.OAuth2Callback {
        public OAuth2Callback(OAuth2Provider provider, String accessToken, String refreshToken) {
            super(provider, accessToken, refreshToken);
        }

        public HTTPClient createApiClient() {
            return new HTTPClient("https://discord.com/").bearer(accessToken);
        }

        public OAuth2Profile getProfile() {
            OAuth2Profile profile = new OAuth2Profile();

            AbstractObject data = createApiClient().get("/api/users/@me")
                            .bearer(accessToken)
                            .header("User-Agent", "JWSPassportClient/1")
                            .data().object();

            if (data.has("id")) {
                profile.id = data.get("id").string();

                if (data.has("avatar"))
                    profile.avatar = "https://cdn.discordapp.com/avatars/"+profile.id+"/"+data.get("avatar").string()+".png";
            }

            if (data.has("username"))
                profile.name = data.get("username").string();

            if (data.has("email"))
                profile.mail = data.get("email").string();

            data.forEach(profile::set);

            return profile;
        }
    }
}

File: backend-kt/src/main/kotlin/de/interaapps/pastefy/auth/oauth/providers/DiscordOAuth2Provider.kt
package de.interaapps.pastefy.auth.oauth.providers

import de.interaapps.pastefy.auth.oauth.*

class DiscordOAuth2Provider(
    private val clientId: String = "",
    private val clientSecret: String = "",
    private val scopes: List<String> = listOf("email", "identify"),
    private val http: OAuthHttpClient? = null,
) : OAuth2Provider {
    override val name = "discord"

    override fun authorizationUrl(callbackUrl: String, state: String): String = authorizationUrl(
        "https://discord.com/api/oauth2/authorize",
        mapOf(
            "response_type" to "code",
            "client_id" to clientId,
            "prompt" to "consent",
            "scope" to scopes.joinToString(" "),
            "redirect_uri" to callbackUrl,
            "state" to state
        ),
    )

    override fun exchangeCode(code: String, callbackUrl: String): OAuth2Tokens = requireHttp().postForm(
        "https://discord.com/api/oauth2/token",
        mapOf(
            "client_id" to clientId,
            "client_secret" to clientSecret,
            "code" to code,
            "grant_type" to "authorization_code",
            "redirect_uri" to callbackUrl,
            "scope" to scopes.joinToString(" ")
        ),
    ).tokens()

    override fun loadProfile(tokens: OAuth2Tokens): OAuth2Profile {
        val data = requireHttp().get(
            "https://discord.com/api/users/@me",
            mapOf("Authorization" to "Bearer ${tokens.accessToken}")
        )
        val id = data.requiredText("id")
        val avatar = data.optionalText("avatar")?.let { "https://cdn.discordapp.com/avatars/$id/$it.png" }
        return OAuth2Profile(id, data.requiredText("username"), data.optionalText("email"), avatar)
    }

    private fun requireHttp() = requireNotNull(http) { "OAuth HTTP client is required" }
}
